File: src/main/java/club/xiaojiawei/strategy/phase/GameOverPhaseStrategy.java
package club.xiaojiawei.strategy.phase;

import club.xiaojiawei.bean.entity.ExtraEntity;
import club.xiaojiawei.bean.entity.TagChangeEntity;
import club.xiaojiawei.status.War;
import club.xiaojiawei.strategy.AbstractPhaseStrategy;
import club.xiaojiawei.utils.SystemUtil;
import com.sun.jna.platform.win32.User32;
import com.sun.jna.platform.win32.WinDef;
import com.sun.jna.platform.win32.WinUser;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.RandomAccessFile;

/**
 * 游戏结束阶段
 * @author 肖嘉威
 * @date 2022/11/27 13:44
 */
@Slf4j
@Component
public class GameOverPhaseStrategy extends AbstractPhaseStrategy{

    @Override
    protected boolean dealTagChangeThenIsOver(String line, TagChangeEntity tagChangeEntity) {
        over();
        return true;
    }

    @Override
    protected boolean dealShowEntityThenIsOver(String line, ExtraEntity extraEntity) {
        over();
        return true;
    }

    @Override
    protected boolean dealFullEntityThenIsOver(String line, ExtraEntity extraEntity) {
        over();
        return true;
    }

    @Override
    protected boolean dealChangeEntityThenIsOver(String line, ExtraEntity extraEntity) {
        over();
        return true;
    }

    @Override
    protected boolean dealOtherThenIsOver(String line) {
        over();
        return true;
    }

    private void over(){
        WinDef.HWND platformHWND = SystemUtil.findPlatformHWND();
        if (platformHWND != null){
            User32.INSTANCE.ShowWindow(platformHWND, WinUser.SW_MINIMIZE);
        }
        War.setMyTurn(false);
        SystemUtil.stopAllThread();
        War.increaseWarCount();
        try {
            SystemUtil.delay(1000);
            RandomAccessFile accessFile = powerLogListener.getAccessFile();
            accessFile.seek(accessFile.length());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        gameUtil.clickGameEndPageTask();
    }
}

File: Hearthstone-Script/src/main/java/club/xiaojiawei/hsscript/strategy/phase/GameOverPhaseStrategy.kt
package club.xiaojiawei.hsscript.strategy.phase

import club.xiaojiawei.hsscript.bean.log.ExtraEntity
import club.xiaojiawei.hsscript.bean.log.TagChangeEntity
import club.xiaojiawei.hsscript.bean.single.WarEx
import club.xiaojiawei.hsscript.listener.log.PowerLogListener
import club.xiaojiawei.hsscript.strategy.AbstractPhaseStrategy
import club.xiaojiawei.hsscript.utils.GameUtil.addGameEndTask
import club.xiaojiawei.hsscript.utils.SystemUtil
import club.xiaojiawei.status.War.isMyTurn
import java.io.IOException

/**
 * 游戏结束阶段
 * @author 肖嘉威
 * @date 2022/11/27 13:44
 */
object GameOverPhaseStrategy : AbstractPhaseStrategy() {
    override fun dealTagChangeThenIsOver(line: String, tagChangeEntity: TagChangeEntity): Boolean {
        over()
        return true
    }

    override fun dealShowEntityThenIsOver(line: String, extraEntity: ExtraEntity): Boolean {
        over()
        return true
    }

    override fun dealFullEntityThenIsOver(line: String, extraEntity: ExtraEntity): Boolean {
        over()
        return true
    }

    override fun dealChangeEntityThenIsOver(line: String, extraEntity: ExtraEntity): Boolean {
        over()
        return true
    }

    override fun dealOtherThenIsOver(line: String): Boolean {
        over()
        return true
    }

    private fun over() {
        isMyTurn = false
        cancelAllTask()
        WarEx.endWar()
        try {
            SystemUtil.delay(1000)
            val accessFile = PowerLogListener.logFile
            accessFile?.seek(accessFile.length())
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        addGameEndTask()
        WarEx.reset()
    }
}
